src/python/WMCore/MicroService/DataStructs/MSOutputTemplate.py

```diff
@@ -0,0 +1,266 @@
+    database) documents. And it must have all the methods needed to set or update
+    various fields in the object. So that we assure the uniformity of all documents.
+
+    Template format:
+    {
+    "RequestName": "ReqName",
+    "RequestStatus": "(clompletd|cloed-out|announced),
+    'Campaign': "Campaign",
+    "creationTime": integer timestamp,
+    "lastUpdate": integer timestamp,
+    "isRelVal": (True|False),
+    "isTaken": (True|False),
+    "destination": ["list of locations"],
+    "OutputDatasets": ["list of datasets"],
+    "destinationOutputMap": [{"destination": ["list of locations"],
+                              "datasets": ["list of datasets"]},
+                             {"destination": ["list of locations"],
+                              "datasets": ["list of datasets"]}],
+    "campaignOutputMap": [{"campaignName": "blah",
+                           "datasets": ["list of datasets"]},
+                          {"campaignName": "blah",
+                           "datasets": ["list of datasets"]}],
+    "transferStatus": "blah", # either "pending" or "done",
+    "transferIDs": ["list of transfer IDs"],
+    "numberOfCopies": integer
+    }
+    """
+    # TODO:
+    #    To add an identifier parameter to __init__
+    #    to distinguish inbound from outbound documents.
+
+    # DONE:
+    #    to figure out how to pass params as dict not only kwargs so I can use
+    #    the same logic for both inbound and outbound documents
+
+    # TODO:
+    #    to implement document mutex in a model similar to the one described here:
+    #    https://stackoverflow.com/questions/32728670/mutex-with-mongodb
+    #    ```
+    #    db.collection.update({done: false, taken: false},{$set: {taken: true, takenBy: myIp}});
+    #    ```
+
+    # TODO:
+    #    as soon as any change in the instantiated object happens ( any call of
+    #    a set method) the same change must be updated in Mongodb too -
+    #    this may happen from outside or from the object itself. The former method
+    #    keeps the document independent from the underlying Database technology
+
+    def __init__(self, doc=None, **kwargs):
+        super(MSOutputTemplate, self).__init__(**kwargs)
+        # super(MSOutputTemplate, self).__init__()
+
+        docTemplate = [
+            ('_id', None, unicode),
+            ('RequestName', None, unicode),
+            ('RequestStatus', None, unicode),
+            ('Campaign', None, unicode),
+            ('creationTime', None, Timestamp),
+            ('lastUpdate', None, Timestamp),
+            ('isRelVal', None, bool),
+            ('isTaken', False, bool),
+            ('isTakenby', None, (str, unicode)),
+            ('OutputDatasets', None, list),
+            ('destination', None, list),
+            ('destinationOutputMap', None, dict),
+            ('campaignOutputMap', None, dict),
+            ('transferStatus', None, unicode),
```

Comment:
Ok, this is something I am currently puzzled how to solve. And more precisely because of [1]. I am not feeling quite comfortable of casting things between `str` and `unicode` in the python code before checking them in the template.

[1]
https://api.mongodb.com/python/3.4.0/tutorial.html#a-note-on-unicode-strings